File: 3/ShiftLinkedList.java
class ShiftLinkedList {
	public static void shiftLinkedList(List list, int k) {
		int n=0;
		Node cur = list.head;
		while(cur!= null) {
			n++;
			cur = cur.next;
		}
		
		k = k % n;
		
		if(k == 0) {
			return;
		}
		
		cur = list.head;
		Node newHeadInNext = null;
		for(int i=1; i<n; i++) {
			if(i == n-k) {
				newHeadInNext = cur;
			}
			cur = cur.next;	
		}
		
		cur.next = list.head;
		list.head = newHeadInNext.next;
		newHeadInNext.next = null;
	}

	public static void main(String args[]) {
		List list = new List();
		int k = 2;
		for(int i=0; i<=5; i++) {
			list.add(i);
		}
			
		shiftLinkedList(list, k);
		System.out.println(list);
	}	

}

class List {
	Node head, tail;
	
	@Override
	public String toString() {
		StringBuffer str = new StringBuffer();
		
		Node cur = head;
		while(cur != null) {
			str.append(cur.data);
			str.append(' ');
			cur = cur.next;
		}

		return str.toString();
	}
	

	void add(int data) {
		if(head == null) {
			head = tail = new Node(data);
			return;
		}
			
		tail.next = new Node(data);
		tail = tail.next;
	}

	List(Node head) {
		if(head == null) {
			return;
		}	

		this.head = head;
		Node cur = head;

		while(cur.next != null) {
			cur = cur.next;
		}
		tail = cur;
	}

	List() {
		head = tail = null;
	}
}

class Node {
	int data;
	Node next;

	Node(int data) {
		this.data = data;
	}
}

File: 3/js/src/ShiftLinkedList.ts
import { LinkedList } from "./LinkedList";

function shiftLinkedList(list: LinkedList, k: number) {
	let n = 0;

	let cur = list.head;
	while (cur != null) {
		n++;
		cur = cur.next;
	}

	k = k % n;

	if (n <= 1 || k == 0) {
		return;
	}

	cur = list.head;
	for (let i = 0; i < n - k - 1; i++) {
		cur = cur!.next;
	}

	const newHead = cur!.next;
	cur!.next = null;

	cur = newHead;
	while (cur!.next != null) {
		cur = cur!.next;
	}

	cur!.next = list.head;
	list.head = newHead;
}

(function main() {
	const list = new LinkedList();
	const k = 2;

	for (let i = 0; i <= 5; i++) {
		list.insertEnd(i);
	}

	shiftLinkedList(list, k);
	console.log(list.toString());
})();
